### PR TITLE
Add Kerberos support

### PR DESCRIPTION
**Pull request by srids**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -0,0 +1,77 @@
+=================================
+nodebase topology for clusterdock
```
Comment (dimaspivak) [inline README.rst]: Update name

File: start.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+from clusterdock.utils import wait_for_condition
+
+logger = logging.getLogger('clusterdock.{}'.format(__name__))
+
+DEFAULT_NAMESPACE = 'clusterdock'
+DEFAULT_OPERATING_SYSTEM = 'centos6.6'
+
+KDC_ACL_FILENAME = '/var/kerberos/krb5kdc/kadm5.acl'
+KDC_CONF_FILENAME = '/var/kerberos/krb5kdc/kdc.conf'
+KERBEROS_VOLUME_DIR = '/etc/clusterdock/kerberos'
```
Comment (dimaspivak) [inline start.py]: Nit: move this to end of list to keep all the KDC options together

File: start.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    krb5_conf_data = kdc_node.get_file(KDC_KRB5_CONF_FILENAME)
+    kdc_node.put_file(KDC_KRB5_CONF_FILENAME,
+                      re.sub(r'EXAMPLE.COM', realm,
+                             re.sub(r'example.com', cluster.network,
+                                    re.sub(r'kerberos.example.com', r'{}.{}'.format(kdc_hostname, cluster.network),
+                                           krb5_conf_data))))
+    kdc_conf_data = kdc_node.get_file(KDC_CONF_FILENAME)
+    kdc_node.put_file(KDC_CONF_FILENAME,
+                      re.sub(r'EXAMPLE.COM', realm,
+                             kdc_conf_data.replace(r'[kdcdefaults]',
```
Comment (dimaspivak) [inline start.py]: Why replace over sub here?

File: start.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                                                   '[kdcdefaults]\n max_renewablelife = 7d\n max_life = 1d')))
+    acl_data = kdc_node.get_file(KDC_ACL_FILENAME)
+    kdc_node.put_file(KDC_ACL_FILENAME, re.sub(r'EXAMPLE.COM', realm, acl_data))
+
+    kdc_commands = [
+        'kdb5_util create -s -r {realm} -P kdcadmin'.format(realm=realm),
+        'kadmin.local -q "addprinc -pw {adminpw} admin/admin@{realm}"'.format(adminpw='acladmin', realm=realm)
```
Comment (dimaspivak) [inline start.py]: Nit" adminpw => admin_pw

File: start.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    ]
+
+    logger.info('Starting KDC ...')
+    # Add the following commands before starting kadmin daemon etc.
+    if args.kerberos_principals:
+        principal_list = ['{}@{}'.format(primary, realm) for primary in args.kerberos_principals.split(',')]
```
Comment (dimaspivak) [inline start.py]: If each element of args.kerberos_principals is actually the primary, should we rename the arg?

File: images/centos6.6/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,22 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (dimaspivak) [inline images/centos6.6/Dockerfile]: General comment about these Dockerfiles; can we either document that the "images" referred to in the file name refers only to the images of the KDC (i.e. not the node images) or maybe just change the name of the folder to kdc_images to be more explicit?

File: start.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+        create_keytab_cmd = 'kadmin.local -q "xst -norandkey -k {} {}" '.format(KDC_KEYTAB_FILENAME,
+                                                                                ' '.join(principal_list))
+        kdc_commands.append(create_keytab_cmd)
+
+    kdc_commands.extend([
+        'krb5kdc',
+        'kadmind',
+        'authconfig --enablekrb5 --update',
+        'service sshd start',
```
Comment (dimaspivak) [inline start.py]: How come these services don't automatically start? If the KDC image is based on the centos images, there should be an init system starting them automatically, I'd think?

File: start.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        'service krb5kdc start',
+        'service kadmin start'
+    ])
+
+    # Gather keytab file and krb5.conf file in KERBEROS_VOLUME_DIR directory which is mounted on host.
+    kdc_commands.append('cp {} {}'.format(KDC_KRB5_CONF_FILENAME, KERBEROS_VOLUME_DIR))
+    if args.kerberos_principals:
+        kdc_commands.append('chmod 644 {}'.format(KDC_KEYTAB_FILENAME))
+
+    kdc_node.execute(command="bash -c '{}'".format('; '.join(kdc_commands)), quiet=True)
```
Comment (dimaspivak) [inline start.py]: quiet should be set to `not args.verbose` to allow users to see command output. Same applies to any place we run `execute`

File: start.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+    logger.info('Validating service health ...')
+    _validate_service_health(node=kdc_node, services=['sshd', 'krb5kdc', 'kadmin'])
+
+
+def _validate_service_health(node, services):
+    def condition(node, services):
+        if all('is running' in (node.execute(command='service {} status'.format(service), quiet=True).output)
```
Comment (dimaspivak) [inline start.py]: I think we end up checking whether `'is running'` is in the command status through two separate execute calls, which could lead to race conditions. Perhaps have the condition compose the list of services that don't meet your condition and then use that same list to log which ones aren't set properly? 

Also, perhaps checking exit codes would be a bit more robust than looking for strings in the message?

File: topology.yaml [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    kdc-node:
+        - kdc
+    nodes:
+        - node-1
+        - node-2
+
+start args:
+    --kerberos-config-directory:
+        default: ~/.clusterdock/kerberos
+        help: If specified, mounts this directory to KDC container for Kerberos config files.
```
Comment (dimaspivak) [inline topology.yaml]: Remove `If specified, ` from help message since it has a default.

Comment (dimaspivak) [inline topology.yaml]: Remove periods from end of help messages

File: topology.yaml [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+        metavar: path
+    --kerberos-principals:
```
Comment (dimaspivak) [inline topology.yaml]: As I mentioned earlier, should this actually be `--kerberos-primaries` or `--kerberos-principal-primaries` or something like that?

**Review by dimaspivak — CHANGES_REQUESTED** [review]
Can you run flake8 against your change? I noticed a few nits like line length issues that we should fix before merging (e.g. line length max is 100 in this project).
Done

File: README.rst (hunk 1 above)
Comment (srids) [reply README.rst]: Done

File: start.py (hunk 2 above)
Comment (srids) [reply start.py]: Done

File: start.py (hunk 3 above)
Comment (srids) [reply start.py]: as you can see, its inserting 2 lines by replacing using sub. not sure if it can be done any otherwise.

File: start.py (hunk 4 above)
Comment (srids) [reply start.py]: Done

File: start.py (hunk 5 above)
Comment (srids) [reply start.py]: actually principal is the right naming convention. so changed primary->principal

File: start.py (hunk 7 above)
Comment (srids) [reply start.py]: followed from topology_cdh (0.0.1) but verified that they do start. so took it out.

File: start.py (hunk 8 above)
Comment (srids) [reply start.py]: Done

File: start.py (hunk 9 above)
Comment (srids) [reply start.py]: changed the logic to use exit code

File: images/centos6.6/Dockerfile (hunk 6 above)
Comment (srids) [reply images/centos6.6/Dockerfile]: renamed to kdc_images

File: topology.yaml (hunk 10 above)
Comment (srids) [reply topology.yaml]: Done

File: topology.yaml (hunk 11 above)
Comment (srids) [reply topology.yaml]: principals is the right thing. so sticking to that and took out primary(ies) anywhere else

File: topology.yaml (hunk 10 above)
Comment (srids) [reply topology.yaml]: Done

**Review by dimaspivak — APPROVED** [review]

**Pull request merged**